stl/inc/xlocnum

```diff
@@ -750,366 +748,363 @@ private:
+
         if (_First != _Last) {
-            if (*_First == _Atoms[_Numget_signoff + 1]) { // gather plus sign
+            if (*_First == _Positive_sign) { // gather plus sign
                 *_Ptr++ = '+';
                 ++_First;
-            } else if (*_First == _Atoms[_Numget_signoff]) { // gather minus sign
+            } else if (*_First == _Negative_sign) { // gather minus sign
                 *_Ptr++ = '-';
                 ++_First;
             }
         }
 
-        char* _Leading = _Ptr; // remember backstop
-        *_Ptr++        = '0'; // backstop carries from sticky bit
+        char* const _Leading = _Ptr; // remember backstop
+        *_Ptr++              = '0'; // backstop carries from sticky bit
 
         bool _Seendigit  = false; // seen a digit in input
         int _Significant = 0; // number of significant digits
-        int _Pten        = 0; // power of 10 multiplier
         size_t _Idx;
 
         const int _Max_sig_dig = (*_Phexexp == _ENABLE_V2_BEHAVIOR ? _MAX_SIG_DIG_V2 : _MAX_SIG_DIG_V1);
 
-        const char* _Pg = &_Grouping[0];
-        if (*_Pg == CHAR_MAX || *_Pg <= '\0') {
-            for (; _First != _Last && (_Idx = _Find_elem(_Atoms, *_First)) < 10; _Seendigit = true, (void) ++_First) {
-                if (_Max_sig_dig <= _Significant) { // enough digits, scale by 10 and update _Sticky
-                    ++_Pten;
-                    if (0 < _Idx) {
-                        _Sticky = true;
-                    }
-                } else if (_Idx != 0 || _Significant != 0) { // save a significant digit
-                    *_Ptr++ = _Src[_Idx];
-                    ++_Significant;
-                }
+        bool _Parse_hex            = false;
+        bool _Has_dec_leading_zero = false;
+        if (_First != _Last && *_First == _Zero_ch) {
+            ++_First;
+            if (_First == _Last) { // "0" only
+                *_Ptr = '\0';
+                return 0;
+            }
+            if (*_First == _Atoms[_Numget_xoff] || *_First == _Atoms[_Numget_xoff + 1]) { // 0X or 0x
+                _Parse_hex = true;
+                ++_First; // discard 0X or 0x for further parsing
+            } else {
+                _Has_dec_leading_zero = true;
             }
-        } else { // grouping specified, gather digits and group sizes
-            const _Elem _Kseparator = _Grouping.empty() ? _Elem{} : _Punct_fac.thousands_sep();
-            string _Groups(1, '\0');
-            size_t _Group = 0;
+        }
 
-            for (; _First != _Last; ++_First) {
-                if ((_Idx = _Find_elem(_Atoms, *_First)) < 10) { // got a digit, add to group size
-                    _Seendigit = true;
-                    if (_Max_sig_dig <= _Significant) { // enough digits, scale by 10 and update _Sticky
-                        ++_Pten;
-                        if (0 < _Idx) {
-                            _Sticky = true;
-                        }
+        const char* _Pg = &_Grouping[0];

```

Comment:
```suggestion
        const char* _Current_grouping = &_Grouping[0];
```

maybe?

Reply:
Also, doing this this way is incorrect, I believe. It's assuming that `\0` will always end `_Pg`; however, if we take a look at [\[facet.numpunct.virtuals\]/3]:

> A `string` `vec` used as a vector of integer values, in which each element `vec[i]` represents the number of digits in the group at position i, starting with position 0 as the rightmost group. If `vec.size() <= i`, the number is the same as group `(i - 1)`; **if `(i < 0 || vec[i] <= 0 || vec[i] == CHAR_­MAX)`, the size of the digit group is unlimited.**

(emphasis mine)

This is an existing bug, but could be fixed pretty easily imo.

[\[facet.numpunct.virtuals\]/3]: https://wg21.link/facet.numpunct.virtuals#3

Reply:
It seems that `do_get` for integer types and `do_put` are also affected...

Reply:
I guess it's safe to think `CHAR_MAX` and non-positive value end `grouping` in `do_put`. I'm now changing `_Getffldx`, `_Getffld`, and `_Getifld`.